modules/network/src/main/java/org/apache/ignite/internal/network/DefaultMessagingService.java

```diff
@@ -337,82 +337,167 @@ private List<ClassDescriptorMessage> prepareMarshal(NetworkMessage msg) throws E
+            HandlerContext firstHandlerContext,
+            Iterator<HandlerContext> remainingContexts
+    ) {
+        if (senderIdIsStale(obj)) {
+            logMessageSkipDueToSenderLeft(obj);
+            return;
         }
 
+        unmarshalMessage(obj);
+
         String senderConsistentId = obj.consistentId();
 
         // Unfortunately, since the Messaging Service is used by ScaleCube itself, some messages can be sent
         // before the node is added to the topology. ScaleCubeMessage handler guarantees to handle null sender consistent ID
         // without throwing an exception.
-        assert message instanceof ScaleCubeMessage || senderConsistentId != null;
+        assert payload instanceof ScaleCubeMessage || senderConsistentId != null;
+
+        // If first handler chose an inbound pool, AND some other handlers also want to be executed on an inbound pool, this means that
+        // they will be guaranteed to be executed on the same thread (as the inbound pool is striped, the stripe key is sender+channelId).
+        // Hence these handlers will be executed on the current thread, so, to avoid additional latency, we'll collect them to a list
+        // and execute here directly after executing the first handler, instead of submitting them to the inbound pool.
+        List<NetworkMessageHandler> handlersWantingSameThreadAsFirst = List.of();
 
-        for (NetworkMessageHandler networkMessageHandler : getMessageHandlers(message.groupType())) {
-            networkMessageHandler.onReceived(message, senderConsistentId, correlationId);
+        while (remainingContexts.hasNext()) {
+            HandlerContext handlerContext = remainingContexts.next();
+
+            if (wantSamePool(firstHandlerContext, handlerContext)) {
+                if (handlersWantingSameThreadAsFirst.isEmpty()) {
+                    handlersWantingSameThreadAsFirst = new ArrayList<>();
+                }
+                handlersWantingSameThreadAsFirst.add(handlerContext.handler());
+            } else {
+                Executor executor = chooseExecutorFor(payload, obj, handlerContext.executorChooser());
+                executor.execute(() -> handlerContext.handler().onReceived(payload, senderConsistentId, correlationId));
+            }
+        }
+
+        firstHandlerContext.handler().onReceived(payload, senderConsistentId, correlationId);
+
+        // Now execute those handlers that are guaranteed to be executed on the same thread as the first one.
+        for (NetworkMessageHandler handler : handlersWantingSameThreadAsFirst) {
+            handler.onReceived(payload, senderConsistentId, correlationId);
         }
     }
 
-    private boolean senderIdIsStale(InNetworkObject obj) {
-        return staleIdDetector.isIdStale(obj.launchId());
+    private static boolean wantSamePool(HandlerContext handlerContext1, HandlerContext handlerContext2) {
+        return wantsInboundPool(handlerContext1) && wantsInboundPool(handlerContext2);
```

Comment:
Why is it not calculated by Choosers comparison?
handlerContext1.executorChooser() == handlerContext2.executorChooser()

Reply:
Fixed